src/wms/src/main/java/org/geoserver/wms/MapLayerInfo.java

```diff
@@ -289,6 +301,12 @@ public int getCacheMaxAge() {
 
         Hints hints = new Hints(ResourcePool.REPROJECT, Boolean.valueOf(!skipReproject));
 
+        if (userMapCRSForWFSNG(resource, coordinateReferenceSystem)) {
+            // a hint for wfs-ng featuresource to keep the crs in query
+            // to skip un-necessary re-projection
+            hints.put(ResourcePool.MAP_CRS, coordinateReferenceSystem);
```

Comment:
if maplayer is a WFS-NG layer, look for the GetMap CRS inside the feature type metadata, if found, add it to the hints which will be passed down to the SecureFeatureSource reader.